File: mac/Sources/SallyportVault/DLP.swift
import Foundation

/// Redacts known secret patterns and values injected for the current request.
/// This is a response filter, not a complete DLP system.
public enum DLP {
    static let maskGeneric = "«redacted»"
    static let maskCredential = "«redacted:sallyport-credential»"

    private static let patternSources: [(String, NSRegularExpression.Options)] = [
        (#"bearer\s+[A-Za-z0-9._\-]{8,}"#, .caseInsensitive),
        (#"gh[pousr]_[A-Za-z0-9]{20,}"#, []),                       // GitHub
        (#"glpat-[A-Za-z0-9_\-]{16,}"#, []),                        // GitLab PAT
        (#"AKIA[0-9A-Z]{16}"#, []),                                 // AWS access key id
        (#"(?:sk|rk|pk)_(?:live|test)_[A-Za-z0-9]{16,}"#, []),      // Stripe
        (#"sk-(?:ant-)?[A-Za-z0-9_\-]{20,}"#, []),                  // OpenAI / Anthropic
        (#"xox[baprse]-[A-Za-z0-9\-]{10,}"#, []),                   // Slack
        (#"AIza[0-9A-Za-z_\-]{35}"#, []),                           // Google API key
        (#"-----BEGIN [A-Z ]*PRIVATE KEY-----.*?-----END [A-Z ]*PRIVATE KEY-----"#,
         .dotMatchesLineSeparators),
        (#"eyJ[A-Za-z0-9_\-]{10,}\.[A-Za-z0-9_\-]{10,}\.[A-Za-z0-9_\-]{10,}"#, []), // JWT
    ]

    /// Retains regex compilation errors for redaction instead of failing at static initialization.
    private static let compiledPatterns: Result<[NSRegularExpression], Error> = Result {
        try patternSources.map { source, options in
            try NSRegularExpression(pattern: source, options: options)
        }
    }

    /// Mask known secret shapes; returns the redaction count.
    public static func redact(_ data: Data) -> (Data, Int) {
        redactWith(data, secrets: [])
    }

    /// Masks exact injected values before known generic patterns. The caller owns
    /// and should zeroize `secrets` after this call.
    public static func redactWith(_ data: Data, secrets: [Data]) -> (Data, Int) {
        guard case .success(let patterns) = compiledPatterns else {
            guard !data.isEmpty else { return (data, 0) }
            return (Data(maskGeneric.utf8), 1)
        }
        // Work on the UTF-8 string form (API bodies are text/JSON); a lossy decode
        // of a binary body still redacts the matching bytes.
        var s = String(decoding: data, as: UTF8.self)
        var count = 0

        // Match exact injected credentials first.
        for secret in secrets where !secret.isEmpty {
            let needle = String(decoding: secret, as: UTF8.self)
            guard !needle.isEmpty else { continue }
            var n = 0
            var range = s.startIndex..<s.endIndex
            while let hit = s.range(of: needle, range: range) {
                n += 1
                range = hit.upperBound..<s.endIndex
            }
            if n > 0 {
                count += n
                s = s.replacingOccurrences(of: needle, with: maskCredential)
            }
        }

        // Generic shapes.
        for re in patterns {
            let ns = s as NSString
            let matches = re.matches(in: s, range: NSRange(location: 0, length: ns.length))
            if matches.isEmpty { continue }
            count += matches.count
            // Replace right-to-left so earlier ranges stay valid.
            let mutable = NSMutableString(string: s)
            for m in matches.reversed() {
                mutable.replaceCharacters(in: m.range, with: maskGeneric)
            }
            s = mutable as String
        }

        return (Data(s.utf8), count)
    }
}

File: core/internal/dlp/dlp.go
// Package dlp masks injected values and a fixed set of credential patterns.
// It is not a general DLP engine.
package dlp

import (
	"bytes"
	"regexp"
)

var patternExpressions = []string{
	`(?i)bearer\s+[A-Za-z0-9._\-]{8,}`,
	`gh[pousr]_[A-Za-z0-9]{20,}`,                  // GitHub
	`glpat-[A-Za-z0-9_\-]{16,}`,                   // GitLab PAT
	`AKIA[0-9A-Z]{16}`,                            // AWS access key id
	`(?:sk|rk|pk)_(?:live|test)_[A-Za-z0-9]{16,}`, // Stripe
	`sk-(?:ant-)?[A-Za-z0-9_\-]{20,}`,             // OpenAI / Anthropic
	`xox[baprse]-[A-Za-z0-9\-]{10,}`,              // Slack
	`AIza[0-9A-Za-z_\-]{35}`,                      // Google API key
	`(?s)-----BEGIN [A-Z ]*PRIVATE KEY-----.*?-----END [A-Z ]*PRIVATE KEY-----`,
	`eyJ[A-Za-z0-9_\-]{10,}\.[A-Za-z0-9_\-]{10,}\.[A-Za-z0-9_\-]{10,}`, // JWT
}

var patterns, patternsValid = compilePatterns(patternExpressions)

// Static DLP configuration is a security boundary too. MustCompile would turn
// one malformed future pattern into an init-time process crash. Keep the error
// as state; RedactWith then masks the complete non-empty response fail closed.
func compilePatterns(expressions []string) ([]*regexp.Regexp, bool) {
	compiled := make([]*regexp.Regexp, 0, len(expressions))
	for _, expression := range expressions {
		re, err := regexp.Compile(expression)
		if err != nil {
			return nil, false
		}
		compiled = append(compiled, re)
	}
	return compiled, true
}

const (
	// maskGeneric masks a value matched by one of the known secret shapes.
	maskGeneric = "«redacted»"
	// maskCredential marks an exact injected-value match.
	maskCredential = "«redacted:sallyport-credential»"
)

// Redact masks known secret shapes and returns the redaction count.
func Redact(b []byte) ([]byte, int) {
	return RedactWith(b, nil)
}

// RedactWith masks literal injected values and known credential patterns. The
// caller owns secrets and should clear them after this call.
func RedactWith(b []byte, secrets [][]byte) ([]byte, int) {
	count := 0
	// Replacement functions return new slices and do not modify b.
	out := b
	if !patternsValid {
		if len(out) == 0 {
			return []byte{}, 0
		}
		return []byte(maskGeneric), 1
	}
	// Mask exact values before the generic pattern pass.
	for _, s := range secrets {
		if len(s) == 0 {
			continue
		}
		if n := bytes.Count(out, s); n > 0 {
			count += n
			out = bytes.ReplaceAll(out, s, []byte(maskCredential))
			// Remove matches formed across a replacement boundary.
			for bytes.Contains(out, s) {
				n = bytes.Count(out, s)
				count += n
				out = bytes.ReplaceAll(out, s, nil)
			}
		}
	}
	for _, re := range patterns {
		out = re.ReplaceAllFunc(out, func(m []byte) []byte {
			count++
			return []byte(maskGeneric)
		})
	}
	// Repeat exact-value removal because one replacement can form another match.
	for {
		changed := false
		for _, s := range secrets {
			if len(s) == 0 {
				continue
			}
			if n := bytes.Count(out, s); n > 0 {
				count += n
				out = bytes.ReplaceAll(out, s, nil)
				changed = true
			}
		}
		if !changed {
			break
		}
	}
	return out, count
}
